[BUG]: Test: The generated ptx will not be loadable by the current driver.

### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cuda-python/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Runtime Error

### Component

cuda.core

### Describe the bug

Occasionally in testing the following error is displayed:

```
RuntimeWarning: The CUDA driver version is older than the backend version. The generated ptx will not be loadable by the current driver.
```

This can come from the `cuda_core/examples/jit_lto_fractal.py` test in the "Test win-64 / py3.11, 13.2.0, wheels, rtx4090 (WDDM)" configuration.

[Full log](https://github.com/NVIDIA/cuda-python/actions/runs/23875576811/job/69618831328)

### How to Reproduce

Happens occasionally in CI.

I think the root cause is that the CI runner is in an incorrect state -- perhaps there is a race condition between installing the driver and the runtime that we could fix by determining the version once and installing the same version of both.  

I think it is better to fix that root cause than to make the test skippable.

### Expected behavior

All tests should pass.

### Operating System

_No response_

### nvidia-smi output

_No response_

Here is another flake of the same kind:

https://github.com/NVIDIA/cuda-python/actions/runs/24330602768/job/71036864789?pr=1898

**Test win-64 / py3.11, 13.0.2, wheels, rtx4090 (WDDM)**

```
================================== FAILURES ===================================
______________________ test_example[jit_lto_fractal.py] _______________________

example = 'jit_lto_fractal.py'

    @pytest.mark.parametrize("example", sample_files)
    def test_example(example):
        example_path = os.path.join(samples_path, example)
        has_package_requirements_or_skip(example_path)
    
        system_requirement = SYSTEM_REQUIREMENTS.get(example, lambda: True)
        if not system_requirement():
            pytest.skip(f"Skipping {example} due to unmet system requirement")
    
        process = subprocess.run([sys.executable, example_path], capture_output=True)  # noqa: S603
        if process.returncode != 0:
            if process.stdout:
                print(process.stdout.decode(errors="replace"))
            if process.stderr:
                print(process.stderr.decode(errors="replace"), file=sys.stderr)
>           raise AssertionError(f"`{example}` failed ({process.returncode})")
E           AssertionError: `jit_lto_fractal.py` failed (3221225477)

example    = 'jit_lto_fractal.py'
example_path = 'C:\\actions-runner\\_work\\cuda-python\\cuda-python\\cuda_core\\tests\\example_tests\\..\\..\\examples\\jit_lto_fractal.py'
process    = CompletedProcess(args=['C:\\actions-runner\\_work\\_tool\\Python\\3.11.9\\x64\\python.exe', 'C:\\actions-runner\\_work... than the backend version. The generated ptx will not be loadable by the current driver.\r\n  lib = MockLibrary()\r\n')
system_requirement = <function test_example.<locals>.<lambda> at 0x0000012E77005760>

tests\example_tests\test_basic_examples.py:115: AssertionError
---------------------------- Captured stderr call -----------------------------
C:\actions-runner\_work\cuda-python\cuda-python\cuda_core\tests\example_tests\..\..\examples\jit_lto_fractal.py:271: RuntimeWarning: The CUDA driver version is older than the backend version. The generated ptx will not be loadable by the current driver.

  lib = MockLibrary()

```

Cursor GPT-5.4 Extra High Fast

With small manual edits

___

## Issue 1851 / CI analysis

### Context

This analysis is for the flake captured in:

- `issue1851_flake_2026-04-27T0440Z.txt` (posted in [previous comment](https://github.com/NVIDIA/cuda-python/issues/1851#issuecomment-4324415775))

Issue 1851 already describes an occasional Windows CI failure where `cuda.core`
emits:

```text
RuntimeWarning: The CUDA driver version is older than the backend version. The generated ptx will not be loadable by the current driver.
```

The issue text already suspects a CI runner state problem rather than a
deterministic code bug.

### What the flake log shows

The failing job is:

- `Test win-64 / py3.11, 13.2.1, wheels, rtx4090 (WDDM)`

The same log shows:

- build/toolkit target: `13.2.1`
- installed backend/toolkit wheels from the `13.2.x` line
- `nvidia-smi` output:

```text
| NVIDIA-SMI 581.15                 Driver Version: 581.15         CUDA Version: 13.0     |
```

- `cuda.pathfinder` driver query test in the same run:

```text
INFO test_real_query_driver_cuda_version: driver_cuda_version=13.0 (encoded=13000)
INFO test_real_driver: real driver CUDA version=13.0 (encoded=13000)
```

So this was a genuine mismatch:

- driver CUDA version: `13.0`
- backend/toolkit components: `13.2.x`

This is exactly the condition described by the warning.

### Why the example failed

The failing example test is `cuda_core/tests/example_tests/test_basic_examples.py`
running `cuda_core/examples/jit_lto_fractal.py` in a subprocess.

The harness simply does:

- run the example as a subprocess
- fail on any nonzero return code

It does **not** special-case the "driver older than backend" warning.

Inside `jit_lto_fractal.py`, `MockLibrary.__init__()` immediately compiles the
library-owned code to PTX:

- `Program(...).compile("ptx")`

That means the example is exposed to PTX/backend-vs-driver compatibility right
away, before the user code even gets linked.

The warning is emitted at:

- `lib = MockLibrary()`

which matches the log.

### The overlooked access violation detail

The log does **not** literally print the words "access violation", but it does
show the subprocess exit code:

```text
AssertionError: `jit_lto_fractal.py` failed (3221225477)
```

On Windows, decimal `3221225477` is hex `0xC0000005`, which is the NTSTATUS
code for:

- `STATUS_ACCESS_VIOLATION`

So the important detail is:

- the warning itself did not fail the subprocess
- the subprocess kept going after the warning
- it later terminated with a Windows access violation

That detail appears to have been easy to overlook because the log surfaces the
warning text much more prominently than the numeric return code.

### Why the flake is rare now

I do **not** think the warning itself is especially rare in a mismatched runner
state. What is rare is the remaining visible failure.

Why:

1. Many regular `cuda_core` unit tests already treat this exact warning as a
   skip condition.
2. The example subprocess test does not; it continues execution.
3. Therefore the mismatch may occur more often than the red-flake count
   suggests, but only a smaller set of code paths still turn it into a hard
   failure.

This is consistent with the log, which shows the same warning class elsewhere
in `cuda_core` tests without causing those tests to fail.

### Tracing the CI configuration

The failing job comes from the Windows pull-request matrix in:

- `ci/test-matrix.yml`

specifically the row:

- `py3.11`, `CUDA_VER: 13.2.1`, `LOCAL_CTK: 0`, `GPU: rtx4090`,
  `DRIVER: latest`, `DRIVER_MODE: WDDM`

The Windows wheel workflow is:

- `.github/workflows/test-wheel-windows.yml`

That workflow:

1. selects the runner label using the matrix entry, including
   `DRIVER: latest`
2. then unconditionally runs:

- `ci/tools/install_gpu_driver.ps1`

The important part is that the driver installer script is **not keyed off**
`CUDA_VER`. The workflow passes `GPU_TYPE` and `DRIVER_MODE`, but not the CUDA
line that the job is testing.

### The likely CI setup bug

`ci/tools/install_gpu_driver.ps1` currently hard-codes:

```powershell
# This driver is picked to support Windows 11 & CUDA 13.0
$version = '581.15'
```

That is highly significant.

For the failing `13.2.1` Windows wheel job, the repo currently does the
following:

1. matrix says: test CUDA `13.2.1`
2. workflow installs `13.2.1` backend/toolkit wheels
3. Windows driver installer script hard-codes driver `581.15`
4. `nvidia-smi` then reports CUDA driver version `13.0`

This matches the flake log exactly.

So yes: this strongly suggests **another CI setup bug** beyond the older issue
already addressed in PR #1921.

More specifically, the Windows driver update path appears stale:

- it is pinned to a driver explicitly described in the script as supporting
  CUDA `13.0`
- and it is not selected based on `CUDA_VER`

### Expected driver for the failing job

From the current repo configuration, the driver effectively expected by the
failing job is:

- `581.15`
- which `nvidia-smi` reports as `CUDA Version: 13.0`

That is **not** aligned with the job's `CUDA_VER: 13.2.1`.

So the expected state from the current CI code is already mismatched.

### Conclusion

My current conclusion is:

- This flake is still in the issue 1851 family.
- The warning is real, not spurious.
- The example subprocess then crashes later with Windows exit code
  `3221225477` (`0xC0000005`, access violation).
- There is strong evidence of a distinct CI configuration bug on Windows:
  `ci/tools/install_gpu_driver.ps1` hard-codes a CUDA-13.0-class driver for
  jobs that are supposed to test CUDA `13.2.1`.

### Most likely fix direction

The most likely fix is in the Windows CI driver setup, not in the example
itself:

- make the Windows driver install path select a driver version that matches the
  CUDA line under test, or at least a driver new enough for the highest CUDA
  line used in the workflow
- avoid hard-coding a `13.0`-class driver for `13.2.1` jobs

Applying the skip behavior to the example still makes sense as a robustness
improvement, but it might only be a mitigation, not the root fix.

--- a/cuda_core/tests/example_tests/test_basic_examples.py
+++ b/cuda_core/tests/example_tests/test_basic_examples.py
@@ -13,6 +13,7 @@ import warnings
 import pytest
 
 from cuda.core import Device, ManagedMemoryResource, system
+from cuda.core._program import _can_load_generated_ptx
 
 try:
     from cuda.bindings._test_helpers.pep723 import has_package_requirements_or_skip
@@ -82,6 +83,7 @@ def has_recent_memory_pool_support() -> bool:
 SYSTEM_REQUIREMENTS = {
     "memory_pool_resources.py": has_recent_memory_pool_support,
     "gl_interop_plasma.py": has_display,
+    "jit_lto_fractal.py": _can_load_generated_ptx,
     "pytorch_example.py": lambda: (
         has_compute_capability_9_or_higher() and is_x86_64()
     ),  # PyTorch only provides CUDA support for x86_64

